manuals/qcowmount.1
.Dd July  3, 2026
.Dt QCOWMOUNT 1
.Os
.Sh NAME
.Nm qcowmount
.Nd mounts a QEMU Copy-On-Write (QCOW) image file
.Sh SYNOPSIS
.Nm qcowmount
.Op Fl k Ar keys
.Op Fl p Ar password
.Op Fl X Ar extended_options
.Op Fl hvV
.Ar source
.Ar mount_point
.Sh DESCRIPTION
.Nm qcowmount
is a utility to mount a QEMU Copy-On-Write (QCOW) image file
.Pp
.Nm qcowmount
is part of the
.Nm libqcow
package.
.Nm libqcow
is a library to access the QEMU Copy-On-Write (QCOW) image file format
.Pp
.Ar source
a QEMU Copy-On-Write (QCOW) image file
.Ar mount_point
the directory to serve as mount point
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl k Ar keys
specify the keys formatted in base16
.It Fl p Ar password
specify the password (or passphrase)
.It Fl v
verbose output to stderr, while qcowmount will remain running in the foreground
.It Fl V
print version
.It Fl X Ar extended_options
extended options to pass to sub system
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# qcowmount image1.qcow2 /mnt/qcow
qcowmount 20260626
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr qcowinfo 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libqcow/issues
.Sh COPYRIGHT
Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* pyqcow
  - add function to set the keys
* add backing file support
* add in-image snapshots support

File format:
* reference count support

Tools:
* qcowmount - fuse version
  - fuse multi thread support
* qcowmount - dokan version
  - have it run in background (using CreateProcess)?
  - what about unmount?
  - fix volume size hints
  - create compatible dokan lib

Tests:
* libqcow_file_set_keys
* libqcow_file_set_utf8_password
* libqcow_file_set_utf16_password
* libqcow_get_access_flags functions
* libqcow_get_codepage
* libqcow_set_codepage
* libqcow_check_file_signature and libqcow_check_file_signature_wide
* thread-safety

20130922
* see `git log' for more recent change log
* updated msvscpp files
* removed unused libfvalue

20130916
* worked on libcthreads build support

20130831
* updated dependencies
* bug fixes in libqcow and pyqcow
* added libqcow_io_handle_clear
* worked on tests

20130512
* updated dependencies
* worked on Python bindings
* worked on tests

20130501
* worked on compressed images
* updated msvscpp files
* changed libqcow_file_get_size to libqcow_file_get_media_size
* qcowmount:
  - worked on snapshot support
  - worked on dokan support
* QCOW 2 added support for image with last cluster block smaller
  than cluster block size

20130430
* worked on QCOW 1 support

20130428
* worked on encrypted images
* worked on tests
* worked on QCOW 1 support
* worked on snapshots
* worked on compressed images

20130427
* updated dependencies
* worked on tests

20130422
* updated dependencies
* added python bindings
* added PackageMaker files
* added libqcow_file_get_size

20130330
* updated dependencies
* code clean up

20130310
* 2013 update

20120929
* updated msvscpp files

20120924
* qcowmount added -X option

20120922
* code clean up

20120917
* fix in dpkg files

20120915
* code clean up

20120913
* applied updates
* changes for libcdata update

20120821
* applied updates

20120630
* applied updates

20120522
* fix in dpkg files

20120517
* updated msvscpp files

20120516
* applied updates
* added qcowmount destroy

20120225
* 2012 update
* preparations for handling encrypted and/or compressed cluster blocks

20111012
* update for Cygwin static executable support
* update for libfuse detection

20111009
* updated configure.ac and m4 files
* updated common, libuna, libbfio, libfvalue, libcaes
* implemented info handle
* worked on qcowmount
* updated msvscpp files
* changes to pc and spec file

20110918
* updated configure, array type, list type
* updated libcstring, libsystem, libuna, libbfio, libfdata, libfvalue
* added libcaes

20110304
* updated common, libcstring, libsystem, libuna, libbfio, libfdata, libfvalue
* update for mingw compilation
* added msvscpp files
* qcowinfo: included libintl for Cygwin

20110111
* updated libsystem

20110110
* updated libcstring, libuna, libbfio, libfvalue
* worked on read_buffer
* added seek and read tests

20110109
* updated configure.ac
* synced local libraries
* worked on read_buffer

20110102
* worked on level 1 table support

20101231
* initial version



qcowtools/qcowmount.c
/*
 * Mounts a QEMU Copy-On-Write (QCOW) image file.
 *
 * Copyright (C) 2010-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <memory.h>
#include <system_string.h>
#include <types.h>

#include <stdio.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "mount_dokan.h"
#include "mount_fuse.h"
#include "mount_handle.h"
#include "qcowtools_getopt.h"
#include "qcowtools_i18n.h"
#include "qcowtools_libcerror.h"
#include "qcowtools_libclocale.h"
#include "qcowtools_libcnotify.h"
#include "qcowtools_libqcow.h"
#include "qcowtools_output.h"
#include "qcowtools_signal.h"
#include "qcowtools_unused.h"

mount_handle_t *qcowmount_mount_handle = NULL;
int qcowmount_abort                    = 0;

/* Signal handler for qcowmount
 */
void qcowmount_signal_handler(
      qcowtools_signal_t signal QCOWTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "qcowmount_signal_handler";

	QCOWTOOLS_UNREFERENCED_PARAMETER( signal )

	qcowmount_abort = 1;

	if( qcowmount_mount_handle != NULL )
	{
		if( mount_handle_signal_abort(
		     qcowmount_mount_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal mount handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use qcowmount to mount a QEMU Copy-On-Write (QCOW) image file.";

	static qcowtools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'k', "keys", "specify the keys formatted in base16" },
		{ 'p', "password", "specify the password/passphrase" },
		{ 'v', NULL, "verbose output to stderr, while qcowmount will remain running in the foreground" },
		{ 'V', NULL, "print version" },
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
		{ 'X', "extended_options", "extended options to pass to sub system" },
#endif
		{ 0, "image", "a QEMU Copy-On-Write (QCOW) image file" },
		{ 0, "mount_point", "the directory to serve as mount point" },
	};
	system_character_t options_string[ 32 ];

	libqcow_error_t *error                      = NULL;
	system_character_t *option_keys             = NULL;
	system_character_t *option_password         = NULL;
	const system_character_t *path_prefix       = NULL;
	system_character_t *source                  = NULL;
	char *program                               = "qcowmount";
	system_integer_t option                     = 0;
	size_t path_prefix_size                     = 0;
	int number_of_options                       = (int) ( sizeof( options ) / sizeof( qcowtools_option_t ) );
	int verbose                                 = 0;

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	system_character_t *mount_point             = NULL;
	int result                                  = 0;
#endif

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	struct fuse_operations qcowmount_fuse_operations;

	system_character_t *option_extended_options = NULL;

#if defined( HAVE_LIBFUSE3 )
	/* Need to set this to 1 even if there no arguments, otherwise this causes
	 * fuse: empty argv passed to fuse_session_new()
	 */
	char *fuse_argv[ 2 ]                        = { program, NULL };
	struct fuse_args qcowmount_fuse_arguments   = FUSE_ARGS_INIT(1, fuse_argv);
#else
	struct fuse_args qcowmount_fuse_arguments   = FUSE_ARGS_INIT(0, NULL);
	struct fuse_chan *qcowmount_fuse_channel    = NULL;
#endif
	struct fuse *qcowmount_fuse_handle          = NULL;

#elif defined( HAVE_LIBDOKAN )
	DOKAN_OPERATIONS qcowmount_dokan_operations;
	DOKAN_OPTIONS qcowmount_dokan_options;
#endif

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "qcowtools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( qcowtools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	qcowtools_output_version_fprint(
	 stdout,
	 program );

	if( qcowtools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = qcowtools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				qcowtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				qcowtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'k':
				option_keys = optarg;

				break;

			case (system_integer_t) 'p':
				option_password = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				qcowtools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );

#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
			case (system_integer_t) 'X':
				option_extended_options = optarg;

				break;
#endif
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source image.\n" );

		qcowtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind++ ];

	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing mount point.\n" );

		qcowtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) || defined( HAVE_LIBDOKAN )
	mount_point = argv[ optind ];
#endif
	libcnotify_verbose_set(
	 verbose );
	libqcow_notify_set_stream(
	 stderr,
	 NULL );
	libqcow_notify_set_verbose(
	 verbose );

	if( mount_handle_initialize(
	     &qcowmount_mount_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize mount handle.\n" );

		goto on_error;
	}
	if( option_keys != NULL )
	{
		if( mount_handle_set_keys(
		     qcowmount_mount_handle,
		     option_keys,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set keys.\n" );

			goto on_error;
		}
	}
	if( option_password != NULL )
	{
		if( mount_handle_set_password(
		     qcowmount_mount_handle,
		     option_password,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set password.\n" );

			goto on_error;
		}
	}
#if defined( WINAPI )
	path_prefix = _SYSTEM_STRING( "\\QCOW" );
#else
	path_prefix = _SYSTEM_STRING( "/qcow" );
#endif
	path_prefix_size = 1 + system_string_length(
	                        path_prefix );

	if( mount_handle_set_path_prefix(
	     qcowmount_mount_handle,
	     path_prefix,
	     path_prefix_size,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set path prefix.\n" );

		goto on_error;
	}
	if( mount_handle_open(
	     qcowmount_mount_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source image\n" );

		goto on_error;
	}
	if( mount_handle_is_locked(
	     qcowmount_mount_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to unlock source image\n" );

		goto on_error;
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( option_extended_options != NULL )
	{
#if defined( HAVE_LIBFUSE3 )
		// fuse_opt_add_arg: Assertion `!args->argv || args->allocated' failed.
		qcowmount_fuse_arguments.argc = 0;
		qcowmount_fuse_arguments.argv = NULL;
#endif
		/* This argument is required but ignored
		 */
		if( fuse_opt_add_arg(
		     &qcowmount_fuse_arguments,
		     "" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &qcowmount_fuse_arguments,
		     "-o" ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
		if( fuse_opt_add_arg(
		     &qcowmount_fuse_arguments,
		     option_extended_options ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable add fuse arguments.\n" );

			goto on_error;
		}
	}
	if( memory_set(
	     &qcowmount_fuse_operations,
	     0,
	     sizeof( struct fuse_operations ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear fuse operations.\n" );

		goto on_error;
	}
	qcowmount_fuse_operations.open       = &mount_fuse_open;
	qcowmount_fuse_operations.read       = &mount_fuse_read;
	qcowmount_fuse_operations.release    = &mount_fuse_release;
	qcowmount_fuse_operations.opendir    = &mount_fuse_opendir;
	qcowmount_fuse_operations.readdir    = &mount_fuse_readdir;
	qcowmount_fuse_operations.releasedir = &mount_fuse_releasedir;
	qcowmount_fuse_operations.getattr    = &mount_fuse_getattr;
	qcowmount_fuse_operations.destroy    = &mount_fuse_destroy;

#if defined( HAVE_LIBFUSE3 )
	qcowmount_fuse_handle = fuse_new(
	                         &qcowmount_fuse_arguments,
	                         &qcowmount_fuse_operations,
	                         sizeof( struct fuse_operations ),
	                         qcowmount_mount_handle );

	if( qcowmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
	result = fuse_mount(
	          qcowmount_fuse_handle,
	          mount_point );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to fuse mount file system.\n" );

		goto on_error;
	}
#else
	qcowmount_fuse_channel = fuse_mount(
	                          mount_point,
	                          &qcowmount_fuse_arguments );

	if( qcowmount_fuse_channel == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse channel.\n" );

		goto on_error;
	}
	qcowmount_fuse_handle = fuse_new(
	                         qcowmount_fuse_channel,
	                         &qcowmount_fuse_arguments,
	                         &qcowmount_fuse_operations,
	                         sizeof( struct fuse_operations ),
	                         qcowmount_mount_handle );

	if( qcowmount_fuse_handle == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to create fuse handle.\n" );

		goto on_error;
	}
#endif /* defined( HAVE_LIBFUSE3 ) */

	if( verbose == 0 )
	{
		if( fuse_daemonize(
		     0 ) != 0 )
		{
			fprintf(
			 stderr,
			 "Unable to daemonize fuse.\n" );

			goto on_error;
		}
	}
	result = fuse_loop(
	          qcowmount_fuse_handle );

	if( result != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to run fuse loop.\n" );

		goto on_error;
	}
	fuse_unmount(
	 qcowmount_fuse_handle );

	fuse_destroy(
	 qcowmount_fuse_handle );

	fuse_opt_free_args(
	 &qcowmount_fuse_arguments );

	return( EXIT_SUCCESS );

#elif defined( HAVE_LIBDOKAN )
	if( memory_set(
	     &qcowmount_dokan_operations,
	     0,
	     sizeof( DOKAN_OPERATIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan operations.\n" );

		goto on_error;
	}
	if( memory_set(
	     &qcowmount_dokan_options,
	     0,
	     sizeof( DOKAN_OPTIONS ) ) == NULL )
	{
		fprintf(
		 stderr,
		 "Unable to clear dokan options.\n" );

		goto on_error;
	}
	qcowmount_dokan_options.Version    = DOKAN_VERSION;
	qcowmount_dokan_options.MountPoint = mount_point;

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	qcowmount_dokan_options.SingleThread = TRUE;
#else
	qcowmount_dokan_options.ThreadCount  = 0;
#endif
	if( verbose != 0 )
	{
		qcowmount_dokan_options.Options |= DOKAN_OPTION_STDERR;
#if defined( HAVE_DEBUG_OUTPUT )
		qcowmount_dokan_options.Options |= DOKAN_OPTION_DEBUG;
#endif
	}
/* This will only affect the drive properties
	qcowmount_dokan_options.Options |= DOKAN_OPTION_REMOVABLE;
*/

#if ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 )
	qcowmount_dokan_options.Options |= DOKAN_OPTION_KEEP_ALIVE;

	qcowmount_dokan_operations.CreateFile           = &mount_dokan_CreateFile;
	qcowmount_dokan_operations.OpenDirectory        = &mount_dokan_OpenDirectory;
	qcowmount_dokan_operations.CreateDirectory      = NULL;
	qcowmount_dokan_operations.Cleanup              = NULL;
	qcowmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	qcowmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	qcowmount_dokan_operations.WriteFile            = NULL;
	qcowmount_dokan_operations.FlushFileBuffers     = NULL;
	qcowmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	qcowmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	qcowmount_dokan_operations.FindFilesWithPattern = NULL;
	qcowmount_dokan_operations.SetFileAttributes    = NULL;
	qcowmount_dokan_operations.SetFileTime          = NULL;
	qcowmount_dokan_operations.DeleteFile           = NULL;
	qcowmount_dokan_operations.DeleteDirectory      = NULL;
	qcowmount_dokan_operations.MoveFile             = NULL;
	qcowmount_dokan_operations.SetEndOfFile         = NULL;
	qcowmount_dokan_operations.SetAllocationSize    = NULL;
	qcowmount_dokan_operations.LockFile             = NULL;
	qcowmount_dokan_operations.UnlockFile           = NULL;
	qcowmount_dokan_operations.GetFileSecurity      = NULL;
	qcowmount_dokan_operations.SetFileSecurity      = NULL;
	qcowmount_dokan_operations.GetDiskFreeSpace     = NULL;
	qcowmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	qcowmount_dokan_operations.Unmount              = &mount_dokan_Unmount;

#else
	qcowmount_dokan_operations.ZwCreateFile         = &mount_dokan_ZwCreateFile;
	qcowmount_dokan_operations.Cleanup              = NULL;
	qcowmount_dokan_operations.CloseFile            = &mount_dokan_CloseFile;
	qcowmount_dokan_operations.ReadFile             = &mount_dokan_ReadFile;
	qcowmount_dokan_operations.WriteFile            = NULL;
	qcowmount_dokan_operations.FlushFileBuffers     = NULL;
	qcowmount_dokan_operations.GetFileInformation   = &mount_dokan_GetFileInformation;
	qcowmount_dokan_operations.FindFiles            = &mount_dokan_FindFiles;
	qcowmount_dokan_operations.FindFilesWithPattern = NULL;
	qcowmount_dokan_operations.SetFileAttributes    = NULL;
	qcowmount_dokan_operations.SetFileTime          = NULL;
	qcowmount_dokan_operations.DeleteFile           = NULL;
	qcowmount_dokan_operations.DeleteDirectory      = NULL;
	qcowmount_dokan_operations.MoveFile             = NULL;
	qcowmount_dokan_operations.SetEndOfFile         = NULL;
	qcowmount_dokan_operations.SetAllocationSize    = NULL;
	qcowmount_dokan_operations.LockFile             = NULL;
	qcowmount_dokan_operations.UnlockFile           = NULL;
	qcowmount_dokan_operations.GetFileSecurity      = NULL;
	qcowmount_dokan_operations.SetFileSecurity      = NULL;
	qcowmount_dokan_operations.GetDiskFreeSpace     = NULL;
	qcowmount_dokan_operations.GetVolumeInformation = &mount_dokan_GetVolumeInformation;
	qcowmount_dokan_operations.Unmounted            = NULL;
	qcowmount_dokan_operations.FindStreams          = NULL;
	qcowmount_dokan_operations.Mounted              = NULL;

#endif /* ( DOKAN_VERSION >= 600 ) && ( DOKAN_VERSION < 800 ) */

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanInit();
#endif
	result = DokanMain(
	          &qcowmount_dokan_options,
	          &qcowmount_dokan_operations );

#if DOKAN_MINIMUM_COMPATIBLE_VERSION >= 200
	DokanShutdown();
#endif
	switch( result )
	{
		case DOKAN_SUCCESS:
			break;

		case DOKAN_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: generic error\n" );
			break;

		case DOKAN_DRIVE_LETTER_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: bad drive letter\n" );
			break;

		case DOKAN_DRIVER_INSTALL_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to load driver\n" );
			break;

		case DOKAN_START_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: driver error\n" );
			break;

		case DOKAN_MOUNT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unable to assign drive letter\n" );
			break;

		case DOKAN_MOUNT_POINT_ERROR:
			fprintf(
			 stderr,
			 "Unable to run dokan main: mount point error\n" );
			break;

		default:
			fprintf(
			 stderr,
			 "Unable to run dokan main: unknown error: %d\n",
			 result );
			break;
	}
	return( EXIT_SUCCESS );

#else
	fprintf(
	 stderr,
	 "No sub system to mount QEMU Copy-On-Write (QCOW) image file format.\n" );

	return( EXIT_FAILURE );

#endif /* defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE ) */

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
#if defined( HAVE_LIBFUSE ) || defined( HAVE_LIBFUSE3 ) || defined( HAVE_LIBOSXFUSE )
	if( qcowmount_fuse_handle != NULL )
	{
		fuse_destroy(
		 qcowmount_fuse_handle );
	}
	fuse_opt_free_args(
	 &qcowmount_fuse_arguments );
#endif
	if( qcowmount_mount_handle != NULL )
	{
		mount_handle_free(
		 &qcowmount_mount_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

